appcues/src/main/java/com/appcues/AppcuesFirebaseMessagingService.kt

```diff
@@ -0,0 +1,100 @@
+package com.appcues
+
+import android.Manifest.permission
+import android.annotation.SuppressLint
+import android.app.PendingIntent
+import android.content.Context
+import android.content.Intent
+import android.content.pm.PackageManager
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.net.Uri
+import androidx.core.app.ActivityCompat
+import androidx.core.app.NotificationCompat
+import androidx.core.app.NotificationManagerCompat
+import com.google.firebase.messaging.FirebaseMessagingService
+import com.google.firebase.messaging.RemoteMessage
+import java.io.IOException
+import java.net.URL
+
+public class AppcuesFirebaseMessagingService : FirebaseMessagingService() {
+
+    public companion object {
+
+        // Suppressing MissingPermission for POST_NOTIFICATION since this should be done on customer's end.
+        @SuppressLint("MissingPermission")
+        @JvmStatic
+        public fun handleMessage(context: Context, message: RemoteMessage): Boolean {
+            // check for permission
+            if (ActivityCompat.checkSelfPermission(context, permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED)
+                return false
+
+            val data = try {
+                AppcuesMessagingData(message)
+            } catch (_: IllegalStateException) {
+                return false
+            }
+
+            val bitmap: Bitmap? = data.image?.let {
+                try {
+                    val url = URL(it)
+                    BitmapFactory.decodeStream(url.openConnection().getInputStream())
+                } catch (e: IOException) {
+                    null
+                }
+            }
+
+            // val channel = message.data["appcues_android_channel"]
+            val builder = NotificationCompat.Builder(context, "channel")
+                .setContentTitle(data.title)
+                .setContentText(data.body)
+                .setAutoCancel(true)
+
+            if (bitmap != null) {
+                builder.setStyle(NotificationCompat.BigPictureStyle().bigPicture(bitmap).setSummaryText(data.body))
+            } else {
+                builder.setStyle(NotificationCompat.BigTextStyle().bigText(data.body))
+            }
+
+            if (data.deepLink != null) {
+                val intent = Intent()
+                intent.setAction(Intent.ACTION_VIEW)
+                intent.setData(Uri.parse(data.deepLink))
+                intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
+                val pendingIntent =
+                    PendingIntent.getActivity(context, 0, intent, PendingIntent.FLAG_ONE_SHOT or PendingIntent.FLAG_IMMUTABLE)
+                builder.setContentIntent(pendingIntent)
+            }
+
+            NotificationManagerCompat.from(context).notify(1, builder.build())
+
+            return true
+        }
+
+        @JvmStatic
+        public fun setToken(token: String) {
```

Comment:
similar to handleMessage, if they have a custom service they can also setup the token using this

```kotlin
   override fun onNewToken(token: String) {
        super.onNewToken(token)

        AppcuesFirebaseMessagingService.setToken(token)
    }
```